### PR TITLE
Align KTO with DPO: Add VLM tests

### PR DESCRIPTION
**Pull request by albertvillanova**
Align KTO with DPO: Add VLM tests.

This PR significantly expands the test coverage for vision-language model (VLM) support in the KTOTrainer, introducing new tests for the `DataCollatorForVisionUnpairedPreference` and VLM training flows. The new tests ensure correct handling of multimodal token type IDs, KL completion cycling, and proper error handling for vision/text model mismatches. Additionally, the tests verify that model parameters are updated as expected and that edge cases (such as truncation and configuration options) are robustly handled.

Part of:
- #4786

### Changes

**New Vision-Language Model (VLM) Test Suite:**

* Added the `TestKTOTrainerVLM` class, which covers end-to-end VLM training for a variety of supported models, checks parameter updates, verifies correct behavior with/without KL loss, tests truncation edge cases, and asserts appropriate error handling for misconfigurations and unsupported options.

**Data Collator for Vision Unpaired Preference:**

* Introduced the `TestDataCollatorForVisionUnpairedPreference` class to validate the shape and presence of multimodal token type IDs, output keys for KL and non-KL paths, and correct cycling of KL completions.

---

> [!NOTE]
> **Low Risk**
> Test-only changes with no production code modifications; CI may run heavier vision-dependent jobs when `require_vision` is satisfied.
> 
> **Overview**
> Adds **vision-language model (VLM) test coverage** for experimental `KTOTrainer`, mirroring DPO-style VLM tests (#4786).
> 
> **`TestDataCollatorForVisionUnpairedPreference`** exercises `DataCollatorForVisionUnpairedPreference`: `mm_token_type_ids` / `KL_completion_mm_token_type_ids` must match the width of their paired `input_ids` (transformers ≥5.3.0), required batch keys with and without KL, and KL completion cycling across batch items.
> 
> **`TestKTOTrainerVLM`** adds end-to-end training on `zen-image` for multiple tiny VLMs (version-gated where needed), checks trainable param updates (including LLaVA frozen layers), `apo_zero_unpaired` without KL, text-only data on a VLM (vision weights stay fixed), `max_length` truncation with multimodal tensors, `ValueError` for vision data + text-only model and for `precompute_ref_log_probs` on vision, plus Liger kernel training.
> 
> Test imports now pull in `DataCollatorForVisionUnpairedPreference`, `require_vision`, `transformers`, and `packaging.version` for skip conditions.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit ab9f738aa587d4102ed020c2d8ab591eb5c56243. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: tests/experimental/test_kto_trainer.py [hunk 1]
```diff
@@ -312,3 +424,202 @@ def dummy_compute_metrics(*args, **kwargs):
...
+        assert trainer.state.log_history[-1]["train_loss"] is not None
+
+    def test_train_vlm_with_max_length(self):
+        # Regression test: mm_token_type_ids (and KL_completion_mm_token_type_ids) must be truncated alongside
+        # input_ids when max_length is set, otherwise a shape mismatch crashes the model forward pass.
+        # max_length=37 truncates 1 completion token (total_len=38) while keeping all image tokens (prompt_len=34) safe.
+        dataset = load_dataset("trl-internal-testing/zen-image", "conversational_unpaired_preference", split="train")
+        training_args = KTOConfig(
+            output_dir=self.tmp_dir,
+            max_length=37,
```
Comment (qgallouedec) [inline tests/experimental/test_kto_trainer.py]:
```suggestion
            max_length=37,  # total_len=38, prompt_len=34 — truncates completion, not image tokens
```

File: tests/experimental/test_kto_trainer.py [hunk 2]
```diff
@@ -312,3 +424,202 @@ def dummy_compute_metrics(*args, **kwargs):
         trainer.train()
 
         assert trainer.state.log_history[-2]["eval_test"] == 0.0
+
+
+@require_vision
+class TestKTOTrainerVLM(TrlTestCase):
```
Comment (qgallouedec) [inline tests/experimental/test_kto_trainer.py:570]: we don't include `test_train_vlm_text_only_data` yet?

Comment (albertvillanova) [reply tests/experimental/test_kto_trainer.py:570]: It is already included: L549

File: tests/experimental/test_kto_trainer.py (hunk 1 above)
Comment (albertvillanova) [reply tests/experimental/test_kto_trainer.py]: Done. Indeed, there were still many non-aligned lines. Now I have addressed them all (I think): [0687611](https://github.com/huggingface/trl/pull/6030/commits/068761144b07a983929b0e197497899c5d344f5f)

File: tests/experimental/test_kto_trainer.py [hunk 3]
```diff
@@ -312,3 +314,312 @@ def dummy_compute_metrics(*args, **kwargs):
...
+        # Check that the params have changed
+        for n, param in previous_trainable_params.items():
+            new_param = trainer.model.get_parameter(n)
+            if model_id in (
+                "trl-internal-testing/tiny-LlavaForConditionalGeneration",
+                "trl-internal-testing/tiny-LlavaNextForConditionalGeneration",
+            ) and ("encoder.layers.1" in n or "post_layernorm" in n):
+                assert torch.equal(param, new_param), f"Param {n} expected frozen by LLaVA design, but changed"
+            else:
+                assert not torch.equal(param, new_param), f"Param {n} is not updated"
```
[the remark this answers is not in the document]
Comment (albertvillanova) [reply tests/experimental/test_kto_trainer.py:640]: I removed `max_steps=3` from all tests.

File: tests/experimental/test_kto_trainer.py (hunk 2 above)
Comment (albertvillanova) [reply tests/experimental/test_kto_trainer.py:570]: L530 after moving tests around.

File: tests/experimental/test_kto_trainer.py [hunk 4]
```diff
@@ -448,3 +564,193 @@ def dummy_compute_metrics(*args, **kwargs):
...
+            pytest.param(
+                "trl-internal-testing/tiny-Qwen3_5MoeForConditionalGeneration-3.6",
+                marks=pytest.mark.skipif(
+                    Version(transformers.__version__) < Version("5.2.0"),
+                    reason="Qwen3.5 models were introduced in transformers-5.2.0",
+                ),
+            ),
+        ],
+    )
+    def test_train_vlm(self, model_id):
```
Comment (qgallouedec) [inline tests/experimental/test_kto_trainer.py:612]: idea for follow-up pr: rename `test_train_vlm` (and friends) to `test_train`

**Review by qgallouedec — APPROVED** [review]
